pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+        M_inv_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+        M_inv_matrix.setUp()
+        tao.setGradientNorm(M_inv_matrix)
+
+        if problem.bounds is not None:
+            lbs = []
+            ubs = []
+            assert len(problem.bounds) == len(problem.reduced_functional.controls)
+            for (lb, ub), control in zip(problem.bounds, taoobjective.reduced_functional.controls):
+                if lb is None:
+                    lb = np.finfo(PETSc.ScalarType).min
+                if ub is None:
+                    ub = np.finfo(PETSc.ScalarType).max
+                lbs.append(lb)
+                ubs.append(ub)
+
+            lb_vec = vec_interface.new_petsc()
+            ub_vec = vec_interface.new_petsc()
+            to_petsc(lb_vec, lbs)
+            to_petsc(ub_vec, ubs)
+            tao.setVariableBounds(lb_vec, ub_vec)
+            lb_vec.destroy()
+            ub_vec.destroy()
+
+        options = PETScOptions(f"_pyadjoint__{tao.name}_")
+        options.update(parameters)
+        tao.setOptionsPrefix(options.options_prefix)
+
+        tao.setFromOptions()
+
+        if tao.getType() in {PETSc.TAO.Type.LMVM, PETSc.TAO.Type.BLMVM}:
+            class InitialHessian:
+                pass
+
+            class InitialHessianPreconditioner:
+                def apply(self, pc, x, y):
+                    X = taoobjective.new_M_dual()
+                    from_petsc(x, X)
+                    assert len(taoobjective.reduced_functional.controls) == len(X)
+                    X = tuple(m._ad_convert_type(x, options=convert_options)
+                              for m, x in zip(taoobjective.reduced_functional.controls, X))
+                    to_petsc(y, X)
+
+            B_0_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                                  InitialHessian(), comm=comm)
+            B_0_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+            B_0_matrix.setUp()
+
+            B_0_matrix_pc = PETSc.PC().createPython(InitialHessianPreconditioner(),
+                                                    comm=comm)
+            B_0_matrix_pc.setOperators(B_0_matrix)
+            B_0_matrix_pc.setUp()
+
+            tao.setLMVMH0(B_0_matrix)
+            ksp = tao.getLMVMH0KSP()
+            ksp.setType(PETSc.KSP.Type.PREONLY)
+            ksp.setTolerances(rtol=0.0, atol=0.0, divtol=None, max_it=1)
+            ksp.setPC(B_0_matrix_pc)
+            ksp.setUp()
+        else:
+            B_0_matrix = None
+            B_0_matrix_pc = None
+
+        x = vec_interface.new_petsc()
+        tao.setSolution(x)
+        tao.setUp()
+
+        super().__init__(problem, parameters)
+        self._taoobjective = taoobjective
+        self._vec_interface = vec_interface
+        self._tao = tao
+        self._x = x
+
+        attach_destroy_finalizer(
+            self, tao, H_matrix, M_inv_matrix, B_0_matrix_pc, B_0_matrix, x)
+
+    @property
+    def taoobjective(self):
```

Comment:
PEP8 name should have underscore.

Reply:
Fixed, although I followed `ROLSolver.rolobjective`.